### PR TITLE
include firebase security rules in apps-package

### PR DESCRIPTION
**Pull request by davidsbailey**
## Overview

Firebase security rules can be uploaded to Firebase via `curl` in json format. However the json format is low-level, so Firebase provides the higher-level Bolt language which can be compiled down to json via the `firebase-bolt` npm. Though the initial rules in this PR are simple, we plan to use Bolt to manage the complexity as we more rules.

Because npm is not installed on our production servers, we run the bolt compiler at the time the apps package is built (on staging or test), and include it in the apps-package which is uploaded to S3. Then we upload it to the Firebase servers in a second step during deploy, which doesn't require npm.

The decision to include the generated `rules.json` in the apps-package is somewhat arbitrary. Arguments for doing so include:
- provides a convenient place to include the `firebase-bolt` npm
- avoids having to install the npm globally and/or upload a new entity to S3
- our code which accesses Firebase also lives in `apps`

Although we intend to access Firebase directly from ruby code, these accesses will be done with elevated privileges and will not be subject to the security rules.
## Documentation
- Bolt compiler: https://www.firebase.com/docs/security/bolt/guide.html
- Uploading security rules: https://www.firebase.com/docs/rest/api/ --> "Retrieving and Updating Security and Firebase Rules"
## Testing
- Tested manually by calling grunt commands directly
- Tested manually by commenting out many lines in `aws/build.rake` and running `rake --rakefile aws/build.rake websites` locally
- Added `firebase_name` and `firebase_secret` to staging chef secrets and will watch what happens there closely after merging
## To confirm after merge
- [x] the rules in rules.bolt are compiled and uploaded to the `cdo-v3-staging` firebase instance
- [ ] a separate change only to `rules.bolt` is sufficient to trigger a new apps package build (it looks like this should work)

### CODE REVIEW

**Comment by davidsbailey** [comment]
Hi @wjordan , this implements a strategy (as best I can remember it) which you had proposed some time back for uploading firebase rules on a per-environment basis as part of the deploy process, without needing npm on production machines. Can you please review (tomorrow is fine)? Thanks!

**Comment by wjordan** [comment]
The general approach LGTM, I have two comments on potentially simplifying the integration code for this feature:
- the `upload` process goes from Ruby (`ci_build.rake#firebase_task`) to JavaScript (`Gruntfile.js#uploadFirebaseRules`) to Ruby (`upload_rules.rb`) to shell (`curl`). Is there a reason not to just call `curl` directly from `firebase_task`, eliminating the extra intermediate steps to JavaScript and back again to Ruby, or is running this through the `grunt` hook and external Ruby script useful for some other purpose?
- similarly, the `compile` process goes from JavaScript (`Gruntfile.js#compileFirebaseRules`) to Ruby (`compile_rules.rb`), then back to an npm-packaged script invoked via shell (`firebase-bolt`), which is not only similarly complicated but I would also worry that the Grunt task won't always have `firebase-bolt` in its path unless it's invoked via npm (`npm run build`). Maybe worth checking with @islemaster on how to simplify integration with the `firebase-bolt` script (I would suggest adding a `scripts` entry to `apps/package.json`, but this could be wrong as I'm not up-to-date on how best to run code within our npm-package context).

**Comment by islemaster** [comment]
As Will points out, it looks like you're just using ruby as a scripting language here; within apps, I'd prefer that we use node for scripting, if only for internal consistency.  These resources might be helpful:
- If you wanted to directly convert your ruby scripts to javascript, the closest thing to Ruby backticks is [`child_process.execSync`](https://nodejs.org/api/child_process.html#child_process_child_process_execsync_command_options).
- Use the [`cdo-config` module](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/script/cdo-config.js) to retrieve config options from merged globals.yml and locals.yml.  This system isn't perfect yet, as it won't retrieve computed values from deployment.rb - but if your firebase stuff is provided in the cdo_secrets cookbook it should end up in globals.yml, so this should work.

File: apps/firebase/compile_rules.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+# Compiles firebase rules from rules.yaml to rules.json using bolt.
+# Note: bolt is installed via npm, which is not available on
+# production machines.
+
+if `which firebase-bolt`.empty?
+  puts "Cannot find firebase-bolt. Please run `npm install` from the apps/ directory."
+  exit(1)
+end
+`mkdir -p ../build/package/firebase`
+`firebase-bolt < rules.bolt > ../build/package/firebase/rules.json`
```
Comment (islemaster) [inline apps/firebase/compile_rules.rb]: If `firebase-bolt` is installed via npm and you want to invoke it outside of an npm script, you should use ``npm bin`/firebase-bolt` to use the absolute path to the executable.

**Comment by davidsbailey** [comment]
Thanks @wjordan, the current approach does seem roundabout.  However, for local development, I also need to be able to upload security rules via the command line. After moving the `curl` command into `firebase_task`, would it make sense to expose this via `rake firebase:upload_rules` defined in `aws/firebase.rake`? Please let me know if you recommend a different approach.

**Comment by davidsbailey** [comment]
Note: during development, firebase rules can now be compiled and uploaded via `npm run compile-firebase-rules && rake firebase:upload_rules`

File: apps/package.json [hunk 2]
```diff
@@ -15,14 +15,15 @@
...
+    "build": "MOOC_DEV=1 grunt build; npm run compile-firebase-rules; grunt logBuildTimes",
+    "build:dist": "NODE_ENV=production MOOC_DEV=0 grunt clean build; npm run compile-firebase-rules; grunt logBuildTimes",
     "start": "MOOC_DEV=1 grunt dev; grunt logBuildTimes",
     "start:applab": "MOOC_DEV=1 grunt dev --app=applab; grunt logBuildTimes",
     "start:gamelab": "MOOC_DEV=1 grunt dev --app=gamelab; grunt logBuildTimes",
     "test-audio": "echo \"Open your browser to http://127.0.0.1:8080/test/audio/audio_test.html\" && http-server .",
     "build:log": "grunt logBuildTimes",
-    "build:stats": "node ./script/build-stats.js"
+    "build:stats": "node ./script/build-stats.js",
+    "compile-firebase-rules": "firebase-bolt < ./firebase/rules.bolt > ./build/package/firebase/rules.json"
```
Comment (islemaster) [inline apps/package.json]: This looks correct.

File: apps/package.json [hunk 3]
```diff
@@ -15,14 +15,15 @@
     "test:integration": "grunt integrationTest",
     "test:entry": "grunt karma:entry",
     "clean": "grunt clean",
-    "build": "MOOC_DEV=1 grunt build; grunt logBuildTimes",
-    "build:dist": "NODE_ENV=production MOOC_DEV=0 grunt clean build; grunt logBuildTimes",
+    "build": "MOOC_DEV=1 grunt build; npm run compile-firebase-rules; grunt logBuildTimes",
+    "build:dist": "NODE_ENV=production MOOC_DEV=0 grunt clean build; npm run compile-firebase-rules; grunt logBuildTimes",
```
Comment (islemaster) [inline apps/package.json]: Do you want `compile-firebase-rules` to run every time here, or only when the preceding build step succeeds? (That is, should this be `;` or `&&`?)

Alternatively, you could encode these as separate tasks `postbuild` and `postbuild:dist` (which npm will automatically run after the respective tasks).  Don't worry about ordering with the `logBuildTimes` task - it's only gathering data from the grunt process either way.

**Comment by wjordan** [comment]
The latest approach looks good to me :+1:

**Comment by islemaster** [comment]
LGTM!

**Pull request merged**